File: app/src/main/java/com/mehmetakiftutuncu/muezzin/services/PrayerTimesUpdaterService.java
package com.mehmetakiftutuncu.muezzin.services;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import androidx.annotation.NonNull;

import com.github.mehmetakiftutuncu.toolbelt.Log;
import com.github.mehmetakiftutuncu.toolbelt.Optional;
import com.mehmetakiftutuncu.muezzin.models.Place;
import com.mehmetakiftutuncu.muezzin.models.PrayerTimeReminder;
import com.mehmetakiftutuncu.muezzin.models.PrayerTimesOfDay;
import com.mehmetakiftutuncu.muezzin.utilities.MuezzinAPI;
import com.mehmetakiftutuncu.muezzin.utilities.Pref;
import com.mehmetakiftutuncu.muezzin.widgetproviders.PrayerTimesWidgetBase;

import java.util.List;

/**
 * Created by akif on 23/05/16.
 */
public class PrayerTimesUpdaterService extends IntentService implements MuezzinAPI.OnPrayerTimesDownloadedListener {
    public PrayerTimesUpdaterService() {
        super("PrayerTimesUpdaterService");
    }

    private Optional<Place> maybeCurrentPlace;

    public static void setUpdater(Context context) {
        Intent intent = new Intent(context, PrayerTimesUpdaterService.class);

        context.startService(intent);
    }

    @Override protected void onHandleIntent(Intent intent) {
        if (intent != null) {
            update();
        }
    }

    private void update() {
        maybeCurrentPlace = Pref.Places.getCurrentPlace(this);

        if (maybeCurrentPlace.isEmpty()) {
            return;
        }

        Place place = maybeCurrentPlace.get();

        Log.debug(getClass(), "Updating prayer times for '%s'...", place);

        MuezzinAPI.get().getPrayerTimes(place, this);
    }

    @Override public void onPrayerTimesDownloaded(@NonNull List<PrayerTimesOfDay> prayerTimes) {
        Place place = maybeCurrentPlace.get();

        if (PrayerTimesOfDay.saveAllPrayerTimes(this, place, prayerTimes)) {
            // Updated prayer times and saved them successfully, now try to reschedule prayer time reminders.
            PrayerTimeReminder.reschedulePrayerTimeReminders(this);
            PrayerTimesWidgetBase.updateAllWidgets(this);
        }
    }

    @Override public void onDownloadPrayerTimesFailed(Exception e) {
        Log.error(getClass(), "Failed to download prayer times for place '%s'!", maybeCurrentPlace.get());
    }
}

File: app/src/main/java/com/mehmetakiftutuncu/muezzin/services/PrayerTimesUpdaterService.kt
package com.mehmetakiftutuncu.muezzin.services

import android.app.IntentService
import android.content.Context
import android.content.Intent
import com.github.mehmetakiftutuncu.toolbelt.Log
import com.mehmetakiftutuncu.muezzin.models.PrayerTimeReminder
import com.mehmetakiftutuncu.muezzin.repositories.PrayerTimesOfDayRepository
import com.mehmetakiftutuncu.muezzin.utilities.MuezzinAPI
import com.mehmetakiftutuncu.muezzin.utilities.Pref
import com.mehmetakiftutuncu.muezzin.widget.PrayerTimesWidget

class PrayerTimesUpdaterService: IntentService("PrayerTimesUpdaterService") {
    override fun onHandleIntent(intent: Intent?) {
        intent?.also {
            val ctx: Context = this

            Pref.Places.getCurrentPlace(ctx)?.also { currentPlace ->
                Log.debug(javaClass, "Updating prayer times for '$currentPlace'...")

                MuezzinAPI.getPrayerTimes(currentPlace, { e ->
                    Log.error(javaClass, e, "Failed to download prayer times for place '$currentPlace'!")
                }) { prayerTimes ->
                    val saved = PrayerTimesOfDayRepository.save(ctx, currentPlace, prayerTimes)

                    if (saved) {
                        // Updated prayer times and saved them successfully
                        // Now try to reschedule prayer time reminders.
                        PrayerTimeReminder.rescheduleReminders(ctx)
                        PrayerTimesWidget.updateAllWidgets(ctx)
                    }
                }
            }
        }
    }

    companion object {
        fun setUpdater(ctx: Context) {
            val intent = Intent(ctx, PrayerTimesUpdaterService::class.java)
            ctx.startService(intent)
        }
    }
}